File: client/src/app/shared/models/borrowals-display-dto.ts
import { BaseDto } from "./base-dto";
import { BookWithAuthorsDto } from "./book-with-authors-dto";
import { BorrowalStatus } from "./borrowal-status-enum";
import { UserInfoDto } from "./user-info-dto";

export interface BorrowalsDisplayDto extends BaseDto {
    borrowalDate: string; 
    borrowalDateStr: string;
    
    dueDate: string;
    isDelayed: boolean;
    dueDateStr: string;
  
    borrowalStatus: BorrowalStatus;
    borrowalStatusStr: string;
  
    // Foreign Keys
    bookId: number;
    appUserId: string;
  
    // FK nav properties to display
    book: BookWithAuthorsDto;
    bookDisplayStr: string;
    bookPictureUrl: string;
  
    appUser: UserInfoDto;
    memberName: string;
    memberEmail: string;
  }

File: Core/DTOs/BorrowalsDisplayDto.cs
using Core.Enums;

namespace Core.DTOs
{
    public class BorrowalsDisplayDto : BaseDto
    {
        public required DateOnly BorrowalDate { get; set; }
        public string BorrowalDateStr => BorrowalDate.ToShortDateString();

        public required DateOnly DueDate { get; set; }
        public bool? IsDelayed => BorrowalStatus == BorrowalStatus.Returned ? null : BorrowalStatus == BorrowalStatus.Out && DueDate < DateOnly.FromDateTime(DateTime.Today);
        public string DueDateStr => DueDate.ToShortDateString();

        public required BorrowalStatus BorrowalStatus { get; set; }
        public string BorrowalStatusStr => BorrowalStatus.ToString();


        // Foreign Keys
        public required int BookId { get; set; }
        public required string AppUserId { get; set; }


        // FK nav properties to display
        public required BookWithAuthorsDto Book { get; set; }
        public string BookDisplayStr => $"{Book.BookTitle} by {Book.AuthorListStr} - {Book.BookPublishedDateStr}";
        public string BookPictureUrl => $"{Book.BookPictureUrl}";


        public required UserInfoDto AppUser { get; set; }
        public string MemberName => $"{AppUser.FirstName} {AppUser.LastName}";
        public string MemberEmail => $"{AppUser.Email}";
    }

    public class BorrowalsDisplayListDto
    {
        public IReadOnlyList<BorrowalsDisplayDto> BorrowalsList { get; set; } = new List<BorrowalsDisplayDto>();
        public int Count => BorrowalsList.Count;
    }
}
